src/webgpu/shader/execution/expression/call/builtin/subgroupAdd.spec.ts

```diff
@@ -0,0 +1,546 @@
+        return new Error(`Invocation ${i}, component ${j}: incorrect result
+- expected: ${expected}
+-      got: ${val}`);
+      }
+    }
+  }
+
+  return undefined;
+}
+
+g.test('data_types')
+  .desc(
+    `Tests subgroup addition for valid data types
+
+Tests a simple addition of all 1 values.
+Reductions expect result to be equal to actual subgroup size.
+Exclusice scans expect result to be equal subgroup invocation id.
+  `
+  )
+  .params(u =>
+    u
+      .combine('type', keysOf(kDataTypes))
+      .filter(t => {
+        // Skip vec3 for simplicity
+        const type = kDataTypes[t.type];
+        if (type instanceof VectorType) {
+          return type.width !== 3;
+        }
+        return true;
+      })
+      .beginSubcases()
+      .combine('wgSize', kWGSizes)
+      .combine('operation', ['reduction', 'exclusive-scan'] as const)
+  )
+  .beforeAllSubcases(t => {
+    const features: GPUFeatureName[] = ['subgroups' as GPUFeatureName];
+    const type = kDataTypes[t.params.type];
+    if (type.requiresF16()) {
+      features.push('shader-f16');
+      features.push('subgroups-f16' as GPUFeatureName);
+    }
+    t.selectDeviceOrSkipTestCase(features);
+  })
+  .fn(async t => {
+    const type = kDataTypes[t.params.type];
+    let numEles = 1;
+    if (type instanceof VectorType) {
+      numEles = type.width;
+    }
+    const scalarType = scalarTypeOf(type);
+    let enables = 'enable subgroups;\n';
+    if (type.requiresF16()) {
+      enables += 'enable f16;\nenable subgroups_f16;\n';
+    }
+
+    // Compatibility mode has lower workgroup limits.
+    const wgThreads = t.params.wgSize[0] * t.params.wgSize[1] * t.params.wgSize[2];
+    const {
+      maxComputeInvocationsPerWorkgroup,
+      maxComputeWorkgroupSizeX,
+      maxComputeWorkgroupSizeY,
+      maxComputeWorkgroupSizeZ,
+    } = t.device.limits;
+    t.skipIf(
+      maxComputeInvocationsPerWorkgroup < wgThreads ||
+        maxComputeWorkgroupSizeX < t.params.wgSize[0] ||
+        maxComputeWorkgroupSizeY < t.params.wgSize[1] ||
+        maxComputeWorkgroupSizeZ < t.params.wgSize[2],
+      'Workgroup size too large'
+    );
+
+    const op = t.params.operation === 'reduction' ? 'subgroupAdd' : 'subgroupExclusiveAdd';
+
+    const wgsl = `
+${enables}
+
+@group(0) @binding(0)
+var<storage> inputs : array<${type.toString()}>;
+
+@group(0) @binding(1)
+var<storage, read_write> outputs : array<${type.toString()}>;
+
+struct Metadata {
+  subgroup_size : array<u32, ${wgThreads}>,
+  subgroup_invocation_id : array<u32, ${wgThreads}>,
+}
+
+@group(0) @binding(2)
+var<storage, read_write> metadata : Metadata;
+
+@compute @workgroup_size(${t.params.wgSize[0]}, ${t.params.wgSize[1]}, ${t.params.wgSize[2]})
+fn main(
+  @builtin(local_invocation_index) lid : u32,
+  @builtin(subgroup_invocation_id) id : u32,
+) {
+  // Record the actual subgroup size for this invocation.
+  // Note: subgroup_size builtin is always a power-of-2 and might be larger
```

Comment:
Related to this, I posted comment about needing to clarify the subgroup_size builtin. https://github.com/gpuweb/gpuweb/issues/4306#issuecomment-2276596259